core/capabilities/syncer.go

```diff
@@ -56,26 +71,486 @@ var defaultStreamConfig = p2ptypes.StreamConfig{
+	registry core.CapabilitiesRegistry,
+	dispatcher remotetypes.Dispatcher,
+	lggr logger.Logger,
+	networkSetup HardcodedDonNetworkSetup,
+	relayer contractReaderFactory,
+	registryAddress string,
+) (*registrySyncer, error) {
+	stopCh := make(services.StopChan)
+	ctx, _ := stopCh.NewCtx()
+	reader, err := newRemoteRegistryReader(ctx, relayer, registryAddress)
+	if err != nil {
+		return nil, err
+	}
+
+	return newRegistrySyncer(
+		stopCh,
+		peerWrapper,
+		registry,
+		dispatcher,
+		lggr,
+		networkSetup,
+		reader,
+	), nil
+}
+
+func newRegistrySyncer(
+	stopCh services.StopChan,
+	peerWrapper p2ptypes.PeerWrapper,
+	registry core.CapabilitiesRegistry,
+	dispatcher remotetypes.Dispatcher,
+	lggr logger.Logger,
+	networkSetup HardcodedDonNetworkSetup,
+	reader reader,
+) *registrySyncer {
 	return &registrySyncer{
+		stopCh:       stopCh,
 		peerWrapper:  peerWrapper,
 		registry:     registry,
 		dispatcher:   dispatcher,
 		networkSetup: networkSetup,
 		lggr:         lggr,
+		reader:       reader,
 	}
 }
 
 func (s *registrySyncer) Start(ctx context.Context) error {
-	s.wg.Add(1)
-	go s.launch(context.Background())
+	s.wg.Add(2)
+	go s.launch()
+	go s.syncLoop()
+	return nil
+}
+
+func (s *registrySyncer) syncLoop() {
+	defer s.wg.Done()
+
+	ctx, cancel := s.stopCh.NewCtx()
+	defer cancel()
+
+	ticker := time.NewTicker(defaultTickInterval)
+	defer ticker.Stop()
+
+	// Sync for a first time outside the loop; this means we'll start a remote
+	// sync immediately once spinning up syncLoop, as by default a ticker will
+	// fire for the first time at T+N, where N is the interval.
+	s.lggr.Debug("starting initial sync with remote registry")
+	err := s.sync(ctx)
+	if err != nil {
+		s.lggr.Errorw("failed to sync with remote registry", "error", err)
+	}
+
+	for {
+		select {
+		case <-s.stopCh:
+			return
+		case <-ticker.C:
+			s.lggr.Debug("starting regular sync with the remote registry")
+			err := s.sync(ctx)
+			if err != nil {
+				s.lggr.Errorw("failed to sync with remote registry", "error", err)
+			}
+		}
+	}
+}
+
+func (s *registrySyncer) sync(ctx context.Context) error {
+	readerState, err := s.reader.state(ctx)
+	if err != nil {
+		return fmt.Errorf("failed to sync with remote registry: %w", err)
+	}
+
+	// Let's start by updating the list of Peers
+	// We do this by creating a new entry for each node belonging
+	// to a public DON.
+	// We also add the hardcoded peers determined by the NetworkSetup.
+	allPeers := make(map[ragetypes.PeerID]p2ptypes.StreamConfig)
+	// TODO: Remove this when we're no longer hard-coding
+	// a `networkSetup`.
+	for p, cfg := range s.networkSetup.allPeers {
+		allPeers[p] = cfg
+	}
+
+	publicDONs := []kcr.CapabilityRegistryDONInfo{}
+	for _, d := range readerState.DONs {
+		if !d.IsPublic {
+			continue
+		}
+
+		publicDONs = append(publicDONs, d)
+
+		for _, nid := range d.NodeP2PIds {
+			allPeers[nid] = defaultStreamConfig
+		}
+	}
+	err = s.peerWrapper.GetPeer().UpdateConnections(allPeers)
```

Comment:
Why are we connecting to all peers even when we don't know if they belong to any relevant DONs? I would expect to connect to peers when:

- The node belongs to the Workflow DON and needs to connect to isPublic DONs with capabilities the node doesn't have. 
- The node belongs to isPublic DON and needs to connect to peers from all acceptsWorkflows DONs.

I assume you skipped this for simplicity early on, but I would leave a comment about this if that's the case.

Reply:
I'd put it in a follow-up backlog task and leave a comment here.